### PR TITLE
Global: api gateway + service discovery

### PR DESCRIPTION
**Pull request by asamal**

### CODE REVIEW

File: sa-eureka-server/src/main/java/com/kyriba/conference/eurekaserver/EurekaServerApplication.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.kyriba.conference.eurekaserver;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.cloud.netflix.eureka.server.EnableEurekaServer;
+
+
+@EnableEurekaServer
+@SpringBootApplication
+public class EurekaServerApplication
```
Comment (ilya-a87) [inline sa-eureka-server/src/main/java/com/kyriba/conference/eurekaserver/EurekaServerApplication.java]: not needed, we can use docker with Eureka

Comment (asamal) [reply sa-eureka-server/src/main/java/com/kyriba/conference/eurekaserver/EurekaServerApplication.java]: done

**Review by ilya-a87 — APPROVED** [review]

File: sa-gateway/src/main/java/com/kyriba/conference/gateway/GatewayApplication.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+public class GatewayApplication
+{
+  public static void main(String[] args)
+  {
+    SpringApplication.run(GatewayApplication.class, args);
+  }
+
+
+  @Bean
+  public RouteLocator myRoutes(RouteLocatorBuilder builder)
```
Comment (xpinjection) [inline sa-gateway/src/main/java/com/kyriba/conference/gateway/GatewayApplication.java]: It is better to define rotes in dedicated config class, not in main app class

File: sa-gateway/src/main/java/com/kyriba/conference/gateway/GatewayApplication.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  {
+    return builder.routes()
+        .route(r -> r.path("/api/v1/sponsorship/**")
+            .filters(f -> f
+                .rewritePath("/api/v1/sponsorship/(?<remains>.*)", "/api/v1/${remains}")
+                .setRequestHeader("Content-Type", MediaType.APPLICATION_JSON_UTF8_VALUE)
```
Comment (xpinjection) [inline sa-gateway/src/main/java/com/kyriba/conference/gateway/GatewayApplication.java]: Don't add content related headers to initial request and transform request body, this is risky operation

File: sa-gateway/src/main/java/com/kyriba/conference/gateway/GatewayApplication.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                .modifyRequestBody(String.class, String.class, MediaType.APPLICATION_JSON_UTF8_VALUE,
+                    (serverWebExchange, s) -> Mono.just(s))
+                .hystrix(c -> c.setFallbackUri("forward:/fallback")))
+            .uri("lb://sponsorship-service")
+            .id("sponsorship"))
+        .build();
+  }
+
+
+  @RequestMapping("/fallback")
```
Comment (xpinjection) [inline sa-gateway/src/main/java/com/kyriba/conference/gateway/GatewayApplication.java:22]: Strange fallback

File: sa-gateway/src/test/java/com/kyriba/conference/gateway/GatewayApplicationTest.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
+import static com.github.tomakehurst.wiremock.client.WireMock.get;
+import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
+import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
+
+
+@Disabled("This test doesn't work, because it is not clear how to mock a load balancer")
+@ExtendWith(SpringExtension.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
Comment (xpinjection) [inline sa-gateway/src/test/java/com/kyriba/conference/gateway/GatewayApplicationTest.java]: API gateway is usually tested at system end to end tests level

File: sa-gateway/src/main/java/com/kyriba/conference/gateway/GatewayApplication.java (hunk 3 above)
Comment (asamal) [reply sa-gateway/src/main/java/com/kyriba/conference/gateway/GatewayApplication.java]: I added it to test my service manually in Chrome (it sets MediaType != APPLICATION_JSON_UTF8_VALUE)

**Pull request merged**